pkg/trait/pod.go

```diff
@@ -0,0 +1,96 @@
+/*
+Licensed to the Apache Software Foundation (ASF) under one or more
+contributor license agreements.  See the NOTICE file distributed with
+this work for additional information regarding copyright ownership.
+The ASF licenses this file to You under the Apache License, Version 2.0
+(the "License"); you may not use this file except in compliance with
+the License.  You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trait
+
+import (
+	"fmt"
+	v1 "github.com/apache/camel-k/pkg/apis/camel/v1"
+	"k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// The PodDisruptionBudget trait is used to configure the pdb
+//
+// +camel-k:trait=pod
+type podTrait struct {
+	BaseTrait      `property:",squash"`
+	MaxUnavailable string `property:"max-unavailable" json:"maxUnavailable,omitempty"`
+	MinAvailable   string `property:"min-available" json:"minAvailable,omitempty"`
+}
+
+func newPodTrait() Trait {
+	return &podTrait{
+		BaseTrait: NewBaseTrait("pod", 900),
+	}
+}
+
+func (t *podTrait) Configure(e *Environment) (bool, error) {
+	if t.Enabled != nil && !*t.Enabled {
+		return false, nil
+	}
+
+	if t.MaxUnavailable == "" && t.MinAvailable == "" {
+		return false, nil
+	}
+	if t.MaxUnavailable != "" && t.MinAvailable != "" {
+		return false, fmt.Errorf("both minAvailable and maxUnavailable can't be set simultaneously")
+	}
+
+	return e.IntegrationInPhase(
+		v1.IntegrationPhaseDeploying,
+		v1.IntegrationPhaseRunning,
+	), nil
+}
+
+func (t *podTrait) Apply(e *Environment) error {
+	pdb := t.generatePodDisruptionBudget(e.Integration.Name, e.Integration)
+	e.Resources.Add(pdb)
+
+	return nil
+}
+
+func (t *podTrait) generatePodDisruptionBudget(pdbName string, integration *v1.Integration) *v1beta1.PodDisruptionBudget {
+	spec := v1beta1.PodDisruptionBudgetSpec{
+		Selector: &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				v1.IntegrationLabel: integration.Name,
+			},
+		},
+	}
+
+	var min, max intstr.IntOrString
+
+	if t.MaxUnavailable != "" {
+		max = intstr.Parse(t.MaxUnavailable)
+		spec.MaxUnavailable = &max
+
+	} else {
```

Comment:
Shouldn't it be `else if t.MinAvailable !=""`?

Reply:
there are only two possible states here `t.MinAvailable!=""` or `t.MaxUnavailable!=""` Other are "filtered" in`Configure` method.

Reply:
What about `t.MinAvailable=="" && t.MaxUnavailable==""`, for which state we could skip the creation of the PDB?

Reply:
it's in the Configure method: https://github.com/apache/camel-k/pull/1787/files#diff-25311c8aa9e2f9d799d92bfecbc73af8737c7b416c1820906218a325fcc3f117R48

Reply:
Ah right. As the pod trait will be responsible for more than the PDB, what do you think of having all the conditions related to the PDB grouped, and the PDB optionally added to the resources?

Reply:
Yeah that makes perfect sense :) I guess my implementation was fine if the trait would manage only the PDB